src/app/clusters/dishwasher-alarm-server/dishwasher-alarm-server.cpp

```diff
@@ -0,0 +1,394 @@
+        status = GetLatchValue(endpoint, &latch);
+        if (status != EMBER_ZCL_STATUS_SUCCESS)
+        {
+            ChipLogProgress(Zcl, "Dishwasher Alarm: latch get ep%d fail", endpoint);
+            return status;
+        }
+
+        if (!latch.HasAll(newState))
+        {
+            latch.Set(newState);
+            status = SetLatchValue(endpoint, latch);
+        }
+    }
+
+    // Generate Notify event.
+    BitMask<AlarmMap> becameActive;
+    becameActive.Set(newState).Clear(currentState);
+    BitMask<AlarmMap> becameInactive;
+    becameInactive.Set(currentState).Clear(newState);
+    // This field SHALL be a copy of the Mask attribute when this event was generated.
+    BitMask<AlarmMap> mask;
+
+    status = GetMaskValue(endpoint, &mask);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return status;
+    }
+    SendNotifyEvent(endpoint, becameActive, becameInactive, newState, mask);
+
+    return status;
+}
+
+bool DishwasherAlarmServer::HasAlarmFeature(EndpointId endpoint)
+{
+    uint32_t featureMap = 0;
+    if (Attributes::FeatureMap::Get(endpoint, &featureMap) != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return false;
+    }
+
+    if (featureMap != 0)
+    {
+        return true;
+    }
+    return false;
+}
+
+void DishwasherAlarmServer::SendNotifyEvent(EndpointId endpointId, BitMask<AlarmMap> becameActive, BitMask<AlarmMap> becameInactive,
+                                            BitMask<AlarmMap> newState, BitMask<AlarmMap> mask)
+{
+    Events::Notify::Type event{ .active = becameActive, .inactive = becameInactive, .state = newState, .mask = mask };
+    EventNumber eventNumber;
+    CHIP_ERROR error = LogEvent(event, endpointId, eventNumber);
+    if (CHIP_NO_ERROR != error)
+    {
+        ChipLogError(Zcl, "[Notify] Unable to send notify event: %s [endpointId=%d]", error.AsString(), endpointId);
+    }
+}
+
+static Status modifyEnabledHandler(const app::ConcreteCommandPath & commandPath, const chip::BitMask<AlarmMap> mask)
+{
+    EndpointId endpoint = commandPath.mEndpointId;
+    chip::BitMask<AlarmMap> support;
+
+    if (DishwasherAlarmServer::Instance().GetSupportedValue(endpoint, &support) != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return Status::Failure;
+    }
+
+    // receives this command with a Mask that includes bits that are set for alarms which are not supported
+    if (!support.HasAll(mask))
+    {
+        return Status::InvalidCommand;
+    }
+
+    // A server that is unable to enable a currently suppressed alarm,
+    // or is unable to suppress a currently enabled alarm SHALL respond
+    // with a status code of FAILURE
+    Delegate * delegate = DishwasherAlarm::GetDelegate(endpoint);
+    if (delegate && !(delegate->ModifyEnableAlarmsCallback(mask)))
+    {
+        ChipLogProgress(Zcl, "Unable to modify enabled alarms");
+        return Status::Failure;
+    }
+
+    if (DishwasherAlarmServer::Instance().SetMaskValue(endpoint, mask) != EMBER_ZCL_STATUS_SUCCESS)
```

Comment:
The delegate API should document that the cluster will do this update if ModifyEnabledAlarmsCallback() returns true.